### PR TITLE
SpeechService: improve Humanize() and use it consistently throughout the default personality

### LINKED ISSUE #1927: Humanize() is too verbose and not used everywhere throughout the default personality
## EDDI version in which issue is found
I have observed this since 3.3.0 or even earlier.

## VoiceAttack version in which issue is found (as applicable)
N/A 

## Steps to reproduce
- Perform any transaction with a value in range of hundreds thousands credits.
- Buy or sell a 8A prismatic shield.

## Expected
Expected the verbalization of the transaction value to have a reasonable length.

## Observed
- Sometimes, the verbalization of the value is not shortened enough, resulting in verbally cluttered utterances like "well over seven hundreds sixty four thousands".
- Sometimes (specifically, the module buy/sell events) the total value is not preprocessed at all, resulting in utterances like ""you have just sold a 8A prismatic shield for a total of eight hundreds seventy four millions six hundreds twenty two thousands one hundred and sixty seven credits".

## Investigation
Expect a PR fixing both issues.

### PR DESCRIPTION
**Pull request by intelfx**
The existing implementation of `SpeechService.Humanize()` is generally overly verbose. For instance:

- it does not try to simplify values in range of hundreds thousands (resulting in verbally cluttered utterances such as "well over seven hundreds sixty four thousands"),
- it uses too many descriptor words ("just over", "over", "well over", "nearly", "around") that don't make it any easier to comprehend the speech (esp. for a non-native English speaker) and sometimes even increase the total word count.

Rewrite `SpeechService.Humanize()` to make it shorten numeric values to 1-2 significant digits, keeping the total word limit under 3 (e. g. "over twelve", but not "under seventy five"), not counting the minus sign and the order (i. e. "hundred thousand").

Additionally, use `Humanise()` consistently throughout the default EDDI personality to avoid ridiculous utterances like "you have just sold a 8A prismatic shield for a total of eight hundreds seventy four millions six hundreds twenty two thousands one hundred and sixty seven credits".

Fixes #1927.

### CODE REVIEW

File: SpeechService/Translations.cs [hunk 1]
```diff
@@ -748,109 +757,159 @@ public static string Humanize(decimal? value)
...
 
-            // See if we have an exact match
-            if (((long)(((decimal)value) / (decimal)Math.Pow(10, digits - 1))) * (decimal)(Math.Pow(10, digits - 1)) == value)
+            // emit exact value
+            if (order == "")
             {
-                return minus + number + order;
+                // we recurse only if we know that there _is_ a nested order (i. e. stuff like "51 hundred thousand")
+                // otherwise we won't be able to round correctly
+                Debug.Assert(recursed == false);
```
Comment (Tkael) [inline SpeechService/Translations.cs]: What is the purpose of this?
```cs
                 Debug.Assert(recursed == false); 
```

File: SpeechService/Translations.cs [hunk 2]
```diff
@@ -748,109 +757,159 @@ public static string Humanize(decimal? value)
...
+                return maybeMinus + number;
             }
 
-            // Describe decimal values
-            if (number < 100)
+            // recurse to handle nested orders (i. e. "50 hundred thousand")
```
Comment (Tkael) [inline SpeechService/Translations.cs:816]: Do you mean 5 hundred thousand?

File: SpeechService/Translations.cs [hunk 3]
```diff
@@ -748,109 +757,159 @@ public static string Humanize(decimal? value)
...
+                order = Properties.Phrases.quadrillion;
             }
 
-            // See if we have an exact match
-            if (((long)(((decimal)value) / (decimal)Math.Pow(10, digits - 1))) * (decimal)(Math.Pow(10, digits - 1)) == value)
+            // emit exact value
+            if (order == "")
             {
-                return minus + number + order;
+                // we recurse only if we know that there _is_ a nested order (i. e. stuff like "51 hundred thousand")
```
Comment (Tkael) [inline SpeechService/Translations.cs]: I'm not sure what you mean by `51 hundred thousand`?

File: SpeechService/Translations.cs (hunk 1 above)
Comment (intelfx) [reply SpeechService/Translations.cs]: It's an assertion that I wrote when I was debugging the new implementation to catch a specific logic problem. Should I remove the assertion?

File: SpeechService/Translations.cs (hunk 2 above)
Comment (intelfx) [reply SpeechService/Translations.cs:816]: Does not matter. A "nested order" is, for example, "hundred thousand".

File: SpeechService/Translations.cs (hunk 3 above)
Comment (intelfx) [reply SpeechService/Translations.cs]: See above. If we receive a number in range between `100000` and `1000000`, when we first enter `Humanize()`, we strip the `10^3` and generate a `thousands` suffix, then we recurse to do the same once again to strip the remaining `10^2`.

Same would work with `xxx hundred millions` etc.

File: SpeechService/Translations.cs (hunk 1 above)
Comment (Tkael) [reply SpeechService/Translations.cs]: Yes, it sounds like it'd be helpful to clean up that line and remove `using System.Diagnostics` at the top of the file.

Comment (intelfx) [reply SpeechService/Translations.cs]: Fine.

**Comment by Tkael** [comment]
As should be expected, this PR yields some different results than our current SpeechUnitTests expect. Reproducing the differences here for discussion.
- TestSpeechHumanize1: `Translations.Humanize(-12345)`
  - WAS: "well over minus 12 thousand"
  - IS: "under minus 12 and a half thousand"
- TestSpeechHumanize7: `Translations.Humanize(-51000000)`
  - WAS: "minus 51 million"
  - IS: "over minus 50 million"
- TestSpeechHumanize8: `Translations.Humanize(51000001)`
  - WAS: "just over 51 million"
  - IS: "over 50 million"
- TestSpeechHumanize10: `Translations.Humanize(100000)`
  - WAS: "100 thousand"
  - IS: "1 hundred thousand"
- TestSpeechHumanize14: `Translations.Humanize(-12.1M)`
  - WAS: "over minus 12"
  - IS: "minus 12"
- TestSpeechHumanize15: `Translations.Humanize(-12.01M)`
  - WAS: "just over minus 12"
  - IS: "minus 12"
- TestSpeechHumanize16: `Translations.Humanize(4.36156E14M)`
  - WAS: "just over 436 trillion"
  - IS: "under 4 and a half hundred trillion"
- TestSpeechHumanize17: `Translations.Humanize(9.4571E11M)`
  - WAS: "well over 945 billion"
  - IS: "under 9 and a half hundred billion"
- TestSpeechHumanize18: `Translations.Humanize(9.1235E17M)`
  - WAS: "over 912 quadrillion"
  - IS: "over 9 hundred quadrillion"

**Comment by Tkael** [comment]
My (consolidated) impressions of the above results...
- TestSpeechHumanize1, TestSpeechHumanize7: 
  - ✔️ Over/under is applied a little differently than it was but is still reasonable.
  - ❗❗ New result is actually a little more verbose than the old result.
- TestSpeechHumanize8, TestSpeechHumanize14, TestSpeechHumanize15, TestSpeechHumanize18
  - ✔️✔️✔️✔️ New result is a little less verbose than the old result. 
- TestSpeechHumanize10
  - ✔️ New result is equivalent to the old result. 
- TestSpeechHumanize16, TestSpeechHumanize17
  - ❌❌ New result is oddly constructed.

**Comment by intelfx** [comment]
Right. I completely forgot about the tests, but it's probably for the best because we can now have this discussion.

>> TestSpeechHumanize1: Translations.Humanize(-12345)
>> WAS: "well over minus 12 thousand"
>> IS: "under minus 12 and a half thousand"
>
>> TestSpeechHumanize7: Translations.Humanize(-51000000)
>> WAS: "minus 51 million"
>> IS: "over minus 50 million"
>
> Over/under is applied a little differently than it was but is still reasonable.

True. I'm not sure which one is better. I can invert over/under for negative numbers if you think that'd be better.

> New result is actually a little more verbose than the old result.

True. I'll think what I can do about it.

>> TestSpeechHumanize16: Translations.Humanize(4.36156E14M)
>> WAS: "just over 436 trillion"
>> IS: "under 4 and a half hundred trillion"
>
>> TestSpeechHumanize17: Translations.Humanize(9.4571E11M)
>> WAS: "well over 945 billion"
>> IS: "under 9 and a half hundred billion"
>
> New result is oddly constructed.

I'd say that's by design, at least it sounds quite natural when spoken. How do you think it should look?

**Review by Tkael — CHANGES_REQUESTED** [review]
I'm getting some odd results returned for `TestSpeechHumanize16` and `TestSpeechHumanize17` in `SpeechUnitTests.cs`. Please take another look at those numbers and see if we can make any improvements?

I'm not sure that hundreds should be applied here. It might be cleaner to prefer the un-rounded number from the primary order (436 trillion or 945 billion, respectively).

**Comment by Tkael** [comment]
After having given it some thought, I think the constructions I'd prefer would be "nearly 440 trillion" and "nearly 950 billion"... i.e. if we use the hundreds order then round the tens to the nearest ten (using "over" or "nearly" to indicate the direction of rounding).

**Comment by intelfx** [comment]
OK, I'll do that.
~~I understand this should be a special case for hundreds?~~ Nevermind, I see you've already clarified this:
>  if we use the hundreds <...>

**Comment by Tkael** [comment]
@intelfx I've pushed a change to the default scripts to better utilize `Humanise()` like you recommended here.

For the remainder (Translations.cs), I agree that the `Humanise()` function probably needs further refinement. That said, I have decided to close this PR both because it seems to have become abandoned and because I have some unresolved misgivings about the changes that you're proposing. If you decide to refine and resubmit then we'll certainly take another look at it.

o7 and thank you for taking the time to submit this.

**Comment by intelfx** [comment]
You could have cherry-picked that specific commit to preserve authorship.

As to the Humanise() rework itself, yes, unfortunately I took a break from Elite due to becoming overwhelmed with real life but I expect to get back to it someday. Why close the PR? This serves no purpose but just fragments the discussion. I would kindly ask to reopen both the PR and the issue because they are not resolved.

**Comment by Tkael** [comment]
I apologize. I chose not to cherry pick the commit because I wanted to make sure that we didn't inadvertently overwrite any other changes to the scripts. 

PR re-opened as requested.

**Comment by Tkael** [comment]
Placing on hold pending evaluation of #2012.

**Comment by richardbuckle** [comment]
Subsumed into #2012. Many thanks for motivating s to look at this!

**Pull request closed without merging**